File: 03-系统编程/01-网络编程/IO模型/MSG-OOB紧急数据/go/main.go
package main

import "fmt"

var scCount int
var scFails []string

func ok(cond bool, msg string) {
	scCount++
	if !cond {
		scFails = append(scFails, msg)
	}
}

func eq(got, want int, msg string)    { ok(got == want, msg) }
func eqi(got, want int64, msg string) { ok(got == want, msg) }

func selfcheck() (int, []string) {
	scCount, scFails = 0, nil

	ok(Before(1, 2), "before(1,2)")
	ok(!After(1, 2), "!after(1,2)")
	ok(!Before(1, 1), "!before(1,1)")
	ok(Before(0xFFFFFFFF, 0), "u32 回绕")
	eq(urgValid, 0x100, "TCP_URG_VALID")
	eq(urgNotYet, 0x200, "TCP_URG_NOTYET")
	eq(urgRead, 0x400, "TCP_URG_READ")

	// BSD 默认解释
	a := newUrgSock(100, 100, false, false)
	a.OnSegment(100, 1, []byte("ABC"), 5, 0)
	eqi(a.UrgSeq, 100, "默认解释 ptr=100")
	eq(a.UrgData, urgValid|0x41, "取到 'A'")
	eq(a.Sockatmark(), 1, "在标记处")
	a.RcvNxt = 103
	eqi(a.Inq(), 0, "SIOCINQ 被截断到 0")
	eqi(a.RecvLimit(100, 3), 0, "普通读上限 0")
	c, hole := a.ReadChunk(100, 3)
	eqi(c, 2, "非 inline 读到 2 字节")
	eq(hole, 1, "urg_hole=1")

	// URGINLINE 成对
	b := newUrgSock(100, 100, true, false)
	b.OnSegment(100, 1, []byte("ABC"), 5, 0)
	b.RcvNxt = 103
	eqi(b.Inq(), 3, "inline -> SIOCINQ 3")
	eqi(b.RecvLimit(100, 3), 0, "inline 也截断")
	c, hole = b.ReadChunk(100, 3)
	eqi(c, 3, "inline 读到 3 字节")
	eq(hole, 0, "inline 无洞")

	// RFC 1122 解释
	d := newUrgSock(100, 100, false, true)
	d.OnSegment(100, 1, []byte("ABC"), 5, 0)
	eqi(d.UrgSeq, 101, "stdurg ptr=101")
	eq(d.UrgData, urgValid|0x42, "取到 'B'")

	// 三条守卫
	e := newUrgSock(100, 105, false, false)
	eq(e.OnSegment(100, 1, []byte("ABC"), 5, 0).Result, "ignored_replay", "不回放")
	f := newUrgSock(100, 100, false, false)
	f.UrgData = urgNotYet
	f.UrgSeq = 102
	eq(f.OnSegment(100, 3, []byte("ABCDE"), 5, 0).Result, "ignored_duplicate", "旧指针")
	eq(f.OnSegment(100, 4, []byte("ABCDE"), 5, 0).Result, "accepted", "新指针(成对)")
	g := newUrgSock(200, 200, false, false)
	eq(g.OnSegment(100, 1, []byte("ABC"), 5, 0).Result, "ignored_already_read", "已读过")

	// Double Dutch
	h := newUrgSock(100, 103, false, false)
	h.UrgData = urgNotYet
	h.UrgSeq = 100
	eq(h.OnSegment(103, 1, []byte("XY"), 5, 0).Result, "accepted_advance_copied_seq", "推进")
	eqi(h.CopiedSeq, 101, "copied_seq 100->101")
	i := newUrgSock(100, 103, true, false)
	i.UrgData = urgNotYet
	i.UrgSeq = 100
	eq(i.OnSegment(103, 1, []byte("XY"), 5, 0).Result, "accepted", "inline 不推进")
	eqi(i.CopiedSeq, 100, "copied_seq 不动")

	// recv_urg 三态
	j := newUrgSock(100, 103, false, false)
	j.UrgData = urgValid | 0x41
	j.UrgSeq = 100
	ret, fl := j.RecvOob(1, false)
	eq(ret, 1, "读到 1 字节")
	eq(fl, msgOob, "msg_flags 带 MSG_OOB")
	eq(j.UrgData, urgRead, "读后 TCP_URG_READ")
	ret, _ = j.RecvOob(1, false)
	eq(ret, -eInval, "再读 -> EINVAL")
	k := newUrgSock(100, 103, false, false)
	k.UrgData = urgValid | 0x41
	k.UrgSeq = 100
	ret, _ = k.RecvOob(1, true)
	eq(ret, 1, "MSG_PEEK 能读")
	eq(k.UrgData, urgValid|0x41, "MSG_PEEK 不改状态")
	ret, fl = k.RecvOob(0, false)
	eq(ret, 0, "len=0 -> 0")
	eq(fl, msgOob|msgTrunc, "len=0 -> MSG_TRUNC")
	l := newUrgSock(100, 103, true, false)
	l.UrgData = urgValid | 0x41
	l.UrgSeq = 100
	ret, _ = l.RecvOob(1, false)
	eq(ret, -eInval, "URGINLINE -> EINVAL")
	n := newUrgSock(100, 103, false, false)
	n.UrgData = urgNotYet
	n.UrgSeq = 100
	ret, _ = n.RecvOob(1, false)
	eq(ret, 0, "NOTYET 返回 0 而非 EINVAL")

	// epoll
	p := newUrgSock(100, 103, false, false)
	p.UrgData = urgValid | 0x41
	p.UrgSeq = 100
	m, target := p.EpollMask(1)
	eq(m, epollPri, "EPOLLPRI")
	eq(target, 2, "rcvlowat +1")
	p.UrgData = urgNotYet
	m, _ = p.EpollMask(1)
	eq(m, 0, "仅 NOTYET 不给 EPOLLPRI")

	return scCount, scFails
}

func main() {
	fmt.Println("== Go 侧自检 ==")
	n, fails := selfcheck()
	fmt.Printf("  assertions=%d  fails=%d\n", n, len(fails))
	for _, f := range fails {
		fmt.Println("  FAIL:", f)
	}

	fmt.Println("\n== 紧急指针两种解释 ==")
	for _, std := range []bool{false, true} {
		s := newUrgSock(100, 100, false, std)
		r := s.OnSegment(100, 1, []byte("ABC"), 5, 0)
		s.RcvNxt = 103
		name := "Linux 默认(BSD)"
		if std {
			name = "tcp_stdurg=1(RFC 1122)"
		}
		fmt.Printf("  %-24s ptr=%-4d byte=%q atmark=%d inq=%d\n",
			name, s.UrgSeq, r.Byte, s.Sockatmark(), s.Inq())
	}
}

File: 03-系统编程/01-网络编程/IO模型/MSG-OOB紧急数据/python/main.py
"""演示入口:TCP 紧急数据的解释分歧、三条守卫与两个容易踩的接口语义。"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from urg_model import (                                   # noqa: E402
    TcpUrgSock, TCP_URG_VALID, TCP_URG_NOTYET, TCP_URG_READ,
)


def show(name, s):
    print("  %-28s urg_seq=%-6d urg_data=0x%04x  atmark=%d  inq=%d"
          % (name, s.urg_seq, s.urg_data, s.sockatmark(), s.inq()))


def main():
    print("== ① 紧急指针的两种解释(urg_ptr=1, seq=100, payload=ABC) ==")
    for name, std in (("Linux 默认(BSD 解释)", False), ("tcp_stdurg=1(RFC 1122)", True)):
        s = TcpUrgSock(copied_seq=100, rcv_nxt=100, stdurg=std)
        r = s.on_segment(seq=100, urg_ptr=1, payload=b"ABC")
        s.rcv_nxt = 103
        print("  %-24s -> ptr=%-5d 取到 %r  result=%s"
              % (name, s.urg_seq, chr(r["byte"]), r["result"]))
        print("  %-24s    atmark=%d  SIOCINQ=%d  普通读上限=%d"
              % ("", s.sockatmark(), s.inq(), s.recv_limit(100, 3)))

    print("\n== ② tcp_check_urg 的三条守卫 ==")
    cases = (
        ("指针指向已收过的字节", dict(copied_seq=100, rcv_nxt=105), 100, 1, b"ABC"),
        ("重复/更旧的指针", None, 100, 3, b"ABCDE"),
        ("指针已被读过", dict(copied_seq=200, rcv_nxt=200), 100, 1, b"ABC"),
    )
    for name, kw, seq, up, pay in cases:
        if kw is None:
            s = TcpUrgSock(copied_seq=100, rcv_nxt=100)
            s.urg_data = TCP_URG_NOTYET
            s.urg_seq = 102
        else:
            s = TcpUrgSock(**kw)
        r = s.on_segment(seq=seq, urg_ptr=up, payload=pay)
        print("  %-22s -> %-24s sigurg=%s"
              % (name, r["result"], r["sigurg"]))
    s = TcpUrgSock(copied_seq=100, rcv_nxt=100)
    s.urg_data = TCP_URG_NOTYET
    s.urg_seq = 102
    print("  %-22s -> %-24s sigurg=%s"
          % ("指针更新一位(对照)", s.on_segment(seq=100, urg_ptr=4,
                                          payload=b"ABCDE")["result"], True))

    print("\n== ③ 'Double Dutch' 修正:copied_seq 会被悄悄 +1 ==")
    for name, inl in (("SO_OOBINLINE 关闭", False), ("SO_OOBINLINE 打开", True)):
        s = TcpUrgSock(copied_seq=100, rcv_nxt=103, oobinline=inl)
        s.urg_data = TCP_URG_NOTYET
        s.urg_seq = 100
        r = s.on_segment(seq=103, urg_ptr=1, payload=b"XY")
        print("  %-22s -> %-32s copied_seq=%d"
              % (name, r["result"], s.copied_seq))

    print("\n== ④ SIOCINQ/FIONREAD 在有紧急数据时被截断到标记处 ==")
    for name, inl in (("默认", False), ("SO_OOBINLINE", True)):
        s = TcpUrgSock(copied_seq=100, rcv_nxt=100, oobinline=inl)
        s.on_segment(seq=100, urg_ptr=1, payload=b"ABC")
        s.rcv_nxt = 103
        print("  %-16s 队列里有 3 字节 -> SIOCINQ 报 %d"
              % (name, s.inq()))

    print("\n== ⑤ MSG_OOB 读取的三态机 ==")
    s = TcpUrgSock(copied_seq=100, rcv_nxt=103)
    s.urg_data = TCP_URG_VALID | 0x41
    s.urg_seq = 100
    print("  初始             0x%04x  recv(MSG_OOB)=%r"
          % (s.urg_data, s.recv_oob()))
    print("  读后             0x%04x  recv(MSG_OOB)=%r  <- EINVAL"
          % (s.urg_data, s.recv_oob()))
    t = TcpUrgSock(copied_seq=100, rcv_nxt=103)
    t.urg_data = TCP_URG_NOTYET
    t.urg_seq = 100
    print("  字节还没到(NOTYET) 0x%04x  recv(MSG_OOB)=%r  <- 不是 EINVAL"
          % (t.urg_data, t.recv_oob()))

    print("\n== ⑥ epoll:EPOLLPRI 只在 TCP_URG_VALID 时给 ==")
    for name, ud in (("无紧急数据", 0), ("仅 NOTYET", TCP_URG_NOTYET),
                     ("VALID", TCP_URG_VALID | 0x41)):
        s = TcpUrgSock(copied_seq=100, rcv_nxt=103)
        s.urg_data = ud
        s.urg_seq = 100
        mask, target = s.epoll_mask(target=1)
        print("  %-12s mask=0x%02x  rcvlowat 目标=%d"
              % (name, mask, target))

    print("\n== ⑦ 一句话结论 ==")
    print("  * 「紧急数据」在 Linux/BSD 默认解释下只有 **1 个字节**。")
    print("  * sockatmark() 返回 1/0 且**不摘除**标记;实现就是 SIOCATMARK。")
    print("  * SIGURG 发给 socket owner(SIOCSPGRP / FIOSETOWN / fcntl F_SETOWN)。")
    print("  * select 报异常条件、poll 报 POLLPRI、epoll 报 EPOLLPRI。")


if __name__ == "__main__":
    main()
